
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="NanoCount estimates transcripts abundance from Oxford Nanopore *direct-RNA sequencing* datasets, using an expectation-maximization approach like RSEM, Kallisto, salmon, etc to handle the uncertainty of multi-mapping reads">
      
      
      
        <meta name="author" content="Adrien Leger">
      
      
        <link rel="canonical" href="https://adrienleger.com/NanoCount/nanocount_io/">
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.2.2, mkdocs-material-7.2.4">
    
    
      
        <title>Input / Output files - NanoCount</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.f7f47774.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.3f5d1f46.min.css">
        
      
    
    
    
      
    
    
    
      <link rel="stylesheet" href="../css/ansi-colours.css">
    
      <link rel="stylesheet" href="../css/jupyter-cells.css">
    
      <link rel="stylesheet" href="../css/pandas-dataframe.css">
    
    
      


    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="none" data-md-color-accent="none">
  
    
    <script>function __prefix(e){return new URL("..",location).pathname+"."+e}function __get(e,t=localStorage){return JSON.parse(t.getItem(__prefix(e)))}</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#nanocount-input-output" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="NanoCount" class="md-header__button md-logo" aria-label="NanoCount" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            NanoCount
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Input / Output files
            
          </span>
        </div>
      </div>
    </div>
    
    
    
    
      <div class="md-header__source">
        
<a href="https://github.com/a-slide/NanoCount/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    a-slide/NanoCount
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="NanoCount" class="md-nav__button md-logo" aria-label="NanoCount" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    NanoCount
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/a-slide/NanoCount/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    a-slide/NanoCount
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../installation/" class="md-nav__link">
        Installation instructions
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_3" type="checkbox" id="__nav_3" checked>
      
      <label class="md-nav__link" for="__nav_3">
        Advanced usage
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Advanced usage" data-md-level="1">
        <label class="md-nav__title" for="__nav_3">
          <span class="md-nav__icon md-icon"></span>
          Advanced usage
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Input / Output files
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Input / Output files
      </a>
      
        
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#input-bam-file" class="md-nav__link">
    Input BAM file
  </a>
  
    <nav class="md-nav" aria-label="Input BAM file">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#output-tsv-file" class="md-nav__link">
    Output TSV file
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#output-bam-file" class="md-nav__link">
    Output BAM file
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../NanoCount_CLI_usage/" class="md-nav__link">
        Command Line interface
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../NanoCount_API_usage/" class="md-nav__link">
        Python API
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../contributing/" class="md-nav__link">
        Contributing
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../code_of_conduct/" class="md-nav__link">
        Code of conduct
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../licence/" class="md-nav__link">
        Licence
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#input-bam-file" class="md-nav__link">
    Input BAM file
  </a>
  
    <nav class="md-nav" aria-label="Input BAM file">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#output-tsv-file" class="md-nav__link">
    Output TSV file
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#output-bam-file" class="md-nav__link">
    Output BAM file
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/a-slide/NanoCount/edit/master/docs/nanocount_io.md" title="Edit this page" class="md-content__button md-icon">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
                  </a>
                
                
                <h1 id="nanocount-input-output">NanoCount Input / Output</h1>
<h2 id="input-bam-file">Input BAM file</h2>
<p>NanoCount is meant to be used with Oxford Nanopore <strong>direct-RNA</strong> sequencing datasets only.</p>
<p>Reads must be aligned a <strong>transcriptome reference</strong> using <strong><a href="https://github.com/lh3/minimap2">minimap2</a></strong> with <code>-p 0 -N 10</code> options to retain up to 10 secondary mappings.
For highly repetitive transcriptomes, this value can even be increased.</p>
<p>Since we use a transcriptome reference the alignment algorithm does not have to be splice aware.</p>
<p>Nanocount does not require reads to be sorted and can take either BAM or SAM format.</p>
<p>Here is an example minimap2 command line with optional conversion to BAM format with samtools:</p>
<pre><code>minimap2 -t 8 -ax map-ont -p 0 -N 10 ./data/yeast_ref.fa.gz ./data/yeast_reads.fastq.gz | samtools view -bh &gt; ./output/aligned_reads.bam
</code></pre>
<h3 id="output-tsv-file">Output TSV file</h3>
<p>NanoCount returns a file containing count data per transcripts.
By default only transcripts with at least one read mapped are included in the report.
This behaviour can be changed to include all transcripts with the option <code>extra_tx_info</code>.</p>
<p>Here is a example tabulated count file returned by NanoCount:</p>
<pre><code>transcript_name raw                  est_count          tpm                transcript_length
YHR174W_mRNA    0.5847481080119605   51228.61224671184  584748.1080119605  1314              
YGR192C_mRNA    0.015286737423038144 1339.2404921575258 15286.737423038145 999               
YGR254W_mRNA    0.011624369387633806 1018.3877533118225 11624.369387633806 1314              
YLR110C_mRNA    0.00945119167199341  827.9999999999986  9451.19167199341   402               
YJR009C_mRNA    0.009088112600958011 796.1913687447295  9088.112600958011  999               
YOL086C_mRNA    0.008275499954342055 724.9999999999987  8275.499954342056  1047              
YKL060C_mRNA    0.006597570998082356 577.9999999999991  6597.570998082356  1080              
YBR118W_mRNA    0.005222125833257228 457.49999999999926 5222.125833257228  1377              
YPR080W_mRNA    0.005222125833257228 457.49999999999926 5222.125833257228  1377    
</code></pre>
<p><strong>Description of fields</strong></p>
<ul>
<li>transcript_name : Transcript name as in source Bam/Sam file.</li>
<li>raw: Raw abundance estimates. The sum of all abundance values is 1.</li>
<li>est_count: Estimated counts obtained by multiplying the raw abundance by the number of primary alignments.</li>
<li>tpm: Estimated counts obtained by multiplying the raw abundance by 1M.</li>
<li>transcript_length: Optional column included with the option <code>extra_tx_info</code>.</li>
</ul>
<p>tpm and estimated counts are not normalised by transcript length as it is usually done with Illumina data.
The reason is that in dRNA-Seq one read is supposed to represent a single transcript molecule starting from the polyA tail, even if the fragment doesn't extend to the 5' end.
If using a custom protocol allowing to sequence from internal RNA fragments (whole RNA tailing, degenerated custom adapter), then the prior is not verified any more.</p>
<h3 id="output-bam-file">Output BAM file</h3>
<p>Optionally, users can choose to dump the alignments selected by Nanocount for the transcripts
estimate step, for QC or visualisation purpose. The alignments are written in the same order
as the source BAM file.</p>
                
              
              
                


              
            </article>
          </div>
        </div>
        
      </main>
      
        
<footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        
        <a href="../installation/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Installation instructions" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Installation instructions
            </div>
          </div>
        </a>
      
      
        
        <a href="../NanoCount_CLI_usage/" class="md-footer__link md-footer__link--next" aria-label="Next: Command Line interface" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Command Line interface
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
          <div class="md-footer-copyright__highlight">
            Copyright Â© 2020 Adrien Leger
          </div>
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
        
      </div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "..", "features": [], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "search": "../assets/javascripts/workers/search.709b4209.min.js", "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.febc23d1.min.js"></script>
      
    
  </body>
</html>